### PR TITLE
Get unique values

### LINKED ISSUE #25: Challenge__Get_Unique_Values
Reads a table (CSV file) and finds all the unique entries in a specific column.  

Example:
Imagine a table of names:  
Name    Age    City  
Alice   25     New York  
Bob     30     Los Angeles  
Alice   35     New York  
Charlie 25     Chicago  

If we want the unique names in the "Name" column, the result will be:  
`{'Alice', 'Bob', 'Charlie'}`

### PR DESCRIPTION
**Pull request by Derekkarungani**
This Python code reads a table (CSV file) and finds all the unique entries in a specific column.

### CODE REVIEW

**Comment by AzzaOmer1** [comment]
# Behavior

## Files
- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name - `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

## Unit Tests
- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has:
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

## Function Docstring
- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (e.g., if it's a number, what's the expected range?)
- The return value is described:
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises`:
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for only one assumption about the argument
- [ ] Include 3 or more (passing!) doctests

## The Function
- [x] The function's name describes its behavior
- [x] The function's name matches the file name
- [x] The function has correct type annotations
- [x] The function is not called in the function file

## Strategy

### Do's
- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts
- [ ] The function's strategy is not described in the documentation
- [ ] Comments explain the strategy, not the implementation
- [ ] The function does not have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation
- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and when it's too restricting
- [x] Variables are named with `snake_case`
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] There are no commented lines of code
- [x] There are no print statements anywhere
- [x] The code includes defensive assertions
  - [x] Defensive assertions include as little logic as possible

File: solutions/get_unique_values.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
+import csv
```
Comment (AzzaOmer1) [inline solutions/get_unique_values.py:11]: there are module header and module docstrig before import csv

File: solutions/tests/test_get_unique_values.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
+import csv  # Import the csv module
```
Comment (AzzaOmer1) [inline solutions/tests/test_get_unique_values.py]: there are Module Docstring before import csv

File: solutions/tests/test_get_unique_values.py [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+import os
+import unittest
+
+from solutions.get_unique_values import get_unique_values
+
+
+class TestGetUniqueValues(unittest.TestCase):
```
Comment (AzzaOmer1) [inline solutions/tests/test_get_unique_values.py:18]: Test class itself doesn't have Class Docstring """ """

File: solutions/tests/test_get_unique_values.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+        os.remove("test_data.csv")
+
+    def test_empty_file(self):
+        """Test with an empty file (with header)."""
+        with open("empty_file.csv", "w") as f:
+            f.write("col1,col2\n")
+        self.assertEqual(get_unique_values("empty_file.csv", "col1"), [])
+        os.remove("empty_file.csv")
+
+    def test_unique_values(self):
```
Comment (AzzaOmer1) [inline solutions/tests/test_get_unique_values.py]: Single Assertion per Test: Most tests have a single assertion. In test_unique_values, there are two assertions.

File: solutions/get_unique_values.py [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+
+def get_unique_values(file_path, column_name):
```
Comment (AzzaOmer1) [inline solutions/get_unique_values.py]: Type Annotations: The function does not explicitly use Python type annotations              def get_unique_values(file_path: str, column_name: str) -> list[str]:

File: solutions/get_unique_values.py [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+        list: Unique values in the specified column.
+
+    Raises:
+        FileNotFoundError: If the file path is invalid.
+        KeyError: If the column name doesn't exist in the CSV.
+    """
+    try:
+        with open(file_path, "r") as file:
+            reader = csv.DictReader(file)
+            if column_name not in reader.fieldnames:
```
Comment (AzzaOmer1) [inline solutions/get_unique_values.py:66]: there are no inline comments explaining the strategy of the function for if

File: solutions/get_unique_values.py [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+                raise KeyError(f"Column '{column_name}' does not exist in the CSV.")
+
+            unique_values = set()
+            for row in reader:
```
Comment (AzzaOmer1) [inline solutions/get_unique_values.py:71]: there are no inline comments explaining the strategy of the function of for loop

File: solutions/get_unique_values.py [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+                unique_values.add(row[column_name])
+
+        return list(unique_values)
+    except FileNotFoundError as e:
```
Comment (AzzaOmer1) [inline solutions/get_unique_values.py:76]: there are no inline comments explaining the except

File: solutions/tests/test_get_unique_values.py [hunk 9]
```diff
@@ -0,0 +1,44 @@
+import csv  # Import the csv module
+import os
+import unittest
+
+from solutions.get_unique_values import get_unique_values
+
+
+class TestGetUniqueValues(unittest.TestCase):
+    def setUp(self):
```
Comment (AzzaOmer1) [inline solutions/tests/test_get_unique_values.py:21]: there is no this assertion per Test:

**Review by AzzaOmer1 — CHANGES_REQUESTED** [review]

File: solutions/get_unique_values.py (hunk 1 above)
Comment (Derekkarungani) [reply solutions/get_unique_values.py:11]: Thanks for reviewing! Made all the changes you requested.

File: solutions/get_unique_values.py [hunk 10]
```diff
@@ -1,7 +1,7 @@
 import csv
```
Comment (AzzaOmer1) [inline solutions/get_unique_values.py:11]: there are not a module header and  a module docstrig as between """   """ before import csv ,

File: solutions/tests/test_get_unique_values.py [hunk 11]
```diff
@@ -1,16 +1,18 @@
-import csv  # Import the csv module
+import csv
```
Comment (AzzaOmer1) [inline solutions/tests/test_get_unique_values.py:11]: There is  no a module docstring in the tests file , as between """ """ befor import csv

File: solutions/tests/test_get_unique_values.py [hunk 12]
```diff
@@ -1,16 +1,18 @@
...
 import os
 import unittest
 
 from solutions.get_unique_values import get_unique_values
 
 
 class TestGetUniqueValues(unittest.TestCase):
+    """Tests for the get_unique_values function."""
+
     def setUp(self):
```
Comment (AzzaOmer1) [inline solutions/tests/test_get_unique_values.py:21]: add assertion for this fun of setup ,because Every unit test has one assertion

**Review by AzzaOmer1 — CHANGES_REQUESTED** [review]
there are 3 things to change it

File: solutions/tests/test_get_unique_values.py (hunk 12 above)
Comment (Derekkarungani) [reply solutions/tests/test_get_unique_values.py:21]: I understand that setUp is primarily for setup, not for testing, so is an assertion really needed there?

File: solutions/get_unique_values.py (hunk 10 above)
Comment (Derekkarungani) [reply solutions/get_unique_values.py:11]: I've added them.

File: solutions/get_unique_values.py [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+        column_name (str): Name of the column to extract unique values.
+
+    Returns:
+        list: Unique values in the specified column.
+
+    Raises:
+        FileNotFoundError: If the file path is invalid.
+        KeyError: If the column name doesn't exist in the CSV.
+    """
+    try:
```
Comment (AzzaOmer1) [inline solutions/get_unique_values.py:62]: there are not  3 or more (passing!) doctests before try.

File: solutions/tests/test_get_unique_values.py (hunk 12 above)
Comment (AzzaOmer1) [reply solutions/tests/test_get_unique_values.py:21]: no ,ok

**Review by AzzaOmer1 — CHANGES_REQUESTED** [review]
there is one issue

File: solutions/get_unique_values.py (hunk 13 above)
Comment (Derekkarungani) [reply solutions/get_unique_values.py:62]: I've added them. Thank you!

File: solutions/get_unique_values.py [hunk 14]
```diff
@@ -0,0 +1,61 @@
...
+
+import csv
+
+
+def get_unique_values(file_path: str, column_name: str) -> list:
+    """
+    Get unique values from a specified column in a CSV file.
+
+    Parameters:
+        file_path (str): Path to the CSV file.
```
Comment (AzzaOmer1) [inline solutions/get_unique_values.py]: Assumptions about the arguments (e.g., valid file paths or expected range of values) are missing

File: solutions/get_unique_values.py [hunk 15]
```diff
@@ -0,0 +1,61 @@
...
+        column_name (str): Name of the column to extract unique values.
```
Comment (AzzaOmer1) [inline solutions/get_unique_values.py]: also here, Other assumptions (e.g., order of values) are not documented

File: solutions/get_unique_values.py [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+
+    Returns:
```
Comment (AzzaOmer1) [inline solutions/get_unique_values.py:24]: Other assumptions (e.g., order of values) are not documented.

File: solutions/get_unique_values.py [hunk 17]
```diff
@@ -0,0 +1,61 @@
+# Module for extracting unique values from CSV file.
+"""
+This module provides the get_unique_values function to extract
```
Comment (AzzaOmer1) [inline solutions/get_unique_values.py]: module docstrings did not completed

File: solutions/tests/test_get_unique_values.py [hunk 18]
```diff
@@ -0,0 +1,54 @@
+"""
+This module contains unit tests for the get_unique_values function
```
Comment (AzzaOmer1) [inline solutions/tests/test_get_unique_values.py:4]: module docstrings did not completed

**Review by AzzaOmer1 — CHANGES_REQUESTED** [review]

File: solutions/tests/test_get_unique_values.py (hunk 18 above)
Comment (Derekkarungani) [reply solutions/tests/test_get_unique_values.py:4]: Made the requested changes.

File: solutions/get_unique_values.py (hunk 13 above)
Comment (AzzaOmer1) [reply solutions/get_unique_values.py:62]: yeah, good examples

File: solutions/get_unique_values.py [hunk 19]
```diff
@@ -0,0 +1,81 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+"""
+This module provides functionality to extract unique values from a CSV file.
+
+Author: Derek Karungani
+Date Created: 2024-12-11
+
+It contains the `get_unique_values` function, which takes a file path and
```
Comment (AzzaOmer1) [inline solutions/get_unique_values.py]: combine with module description

File: solutions/tests/test_get_unique_values.py [hunk 20]
```diff
@@ -0,0 +1,59 @@
...
+
+    def test_file_not_found(self):
+        """Test with a non-existent file."""
+        with self.assertRaises(FileNotFoundError):
+            get_unique_values("nonexistent_file.csv", "col1")
+
+    def test_column_not_in_csv(self):
+        """Test with a column not in the CSV."""
+        with self.assertRaises(KeyError):
+            get_unique_values("test_data.csv", "non_existent_column")
```
Comment (AzzaOmer1) [inline solutions/tests/test_get_unique_values.py:59]: good tests

**Review by AzzaOmer1 — CHANGES_REQUESTED** [review]
complete issues

File: solutions/tests/test_get_unique_values.py (hunk 20 above)
Comment (Derekkarungani) [reply solutions/tests/test_get_unique_values.py:59]: Yours is a really thorough review! :)

**Review by AzzaOmer1 — CHANGES_REQUESTED** [review]
Done  the review

**Review by AzzaOmer1 — APPROVED** [review]
Done the review
